drivers/serial/uart_nrfx_uarte.c

```diff
@@ -251,6 +307,453 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+
+	if (nrf_uarte_event_check(uarte, NRF_UARTE_EVENT_ERROR)) {
+		/* register bitfields maps to the defines in uart.h */
+		error = nrf_uarte_errorsrc_get_and_clear(uarte);
+	}
+
+	return error;
+}
+
+#ifdef CONFIG_UART_ASYNC_API
+
+static int uarte_nrfx_tx(struct device *dev, const u8_t *buf, size_t len,
+			 u32_t timeout)
+{
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+	const struct uarte_nrfx_config *cfg = get_dev_config(dev);
+
+	if (data->async->tx) {
+		return -EBUSY;
+	}
+	data->async->tx = buf;
+	data->async->tx_len = len;
+	data->async->tx_cur_idx = 0;
+	nrf_uarte_tx_buffer_set(cfg->uarte_regs, buf, len);
+	nrf_uarte_event_clear(cfg->uarte_regs, NRF_UARTE_EVENT_TXSTARTED);
+	nrf_uarte_task_trigger(cfg->uarte_regs, NRF_UARTE_TASK_STARTTX);
+	if (cfg->uarte_regs->CONFIG & UARTE_CONFIG_HWFC_Msk) {
+		k_delayed_work_submit(&data->async->tx_timeout_work, timeout);
+	}
+	return 0;
+}
+
+#define INTERRUPTS_MASK							       \
+	(NRF_UARTE_INT_ENDRX_MASK | NRF_UARTE_INT_RXSTARTED_MASK |	       \
+	 NRF_UARTE_INT_ERROR_MASK | NRF_UARTE_INT_ENDTX_MASK |		       \
+	 NRF_UARTE_INT_TXSTOPPED_MASK | NRF_UARTE_INT_RXTO_MASK)
+
+static void user_callback(struct device *dev, struct uart_event *evt)
+{
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+
+	if (data->async->evt_handler) {
+		data->async->evt_handler(evt, data->async->user_data);
+	}
+}
+
+static void rx_timeout(struct device *dev)
+{
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+	u32_t read;
+
+	if (IS_ENABLED(CONFIG_UARTE_NRF_HW_ASYNC) && data->async->hw_async) {
+		read = nrfx_timer_capture(&data->async->timer, 0);
+	} else {
+		read = data->async->rx_cnt.cnt;
+	}
+
+	if (read > data->async->last_rx_byte_cnt) {
+		data->async->last_rx_byte_cnt = read;
+		data->async->rx_timeout_left = data->async->rx_timeout;
+	}
+	if (data->async->last_rx_byte_cnt > data->async->last_rx_evt) {
+		if (data->async->rx_timeout_left
+		    < data->async->rx_timeout_slab) {
+			u32_t len = data->async->last_rx_byte_cnt
+				    - data->async->last_rx_evt;
+			struct uart_event evt = {
+				.type = UART_RX_RDY,
+				.data.rx.buf = data->async->cur_rx,
+				.data.rx.len = len,
+				.data.rx.offset = data->async->rx_offset
+			};
+			data->async->rx_offset += len;
+			data->async->last_rx_evt =
+				data->async->last_rx_byte_cnt;
+			user_callback(dev, &evt);
+			data->async->rx_timeout_left = data->async->rx_timeout;
+		} else {
+			data->async->rx_timeout_left -=
+				data->async->rx_timeout_act;
+		}
+	}
+
+	k_delayed_work_submit(&data->async->rx_timeout_work,
+			      data->async->rx_timeout_slab);
+}
+
+void timer_handler(nrf_timer_event_t event_type, void *p_context) { }
+
+static int uarte_nrfx_init(struct device *dev)
+{
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+	const struct uarte_nrfx_config *cfg = get_dev_config(dev);
+	u32_t ret;
+
+	if (IS_ENABLED(CONFIG_UARTE_NRF_HW_ASYNC) && data->async->hw_async) {
+		ret = nrfx_ppi_channel_alloc(&data->async->rx_cnt.ppi);
```

Comment:
what about dppi? maybe it would be worth extracting dppi/ppi stuff to another file with simple call like `ppi_configure` or `gppi_configure` where g stands for generic :)

Reply:
There are some differences in how PPI and DPPI handle signals, I believe it can be done but it won't be a part of this PR